File: go/config/schema.go
// Package config provides configuration schema definition and runtime client
// for the Smoo AI configuration management platform.
package config

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/invopop/jsonschema"
)

// ConfigTier represents the tier of a configuration value.
type ConfigTier string

const (
	// TierPublic is for public configuration values visible to all.
	TierPublic ConfigTier = "public"
	// TierSecret is for secret configuration values (encrypted at rest).
	TierSecret ConfigTier = "secret"
	// TierFeatureFlag is for feature flag values.
	TierFeatureFlag ConfigTier = "feature_flag"
)

// ConfigDefinition holds the schema definition result from DefineConfig.
type ConfigDefinition struct {
	PublicSchema      map[string]any `json:"public_schema"`
	SecretSchema      map[string]any `json:"secret_schema"`
	FeatureFlagSchema map[string]any `json:"feature_flag_schema"`
	JSONSchema        map[string]any `json:"json_schema"`
}

// DefineConfig creates a configuration definition from optional tier schemas.
// Each schema should be a JSON Schema object describing that tier's configuration.
// Validates each tier's schema for cross-language compatibility and prints
// warnings for unsupported features.
func DefineConfig(publicSchema, secretSchema, featureFlagSchema map[string]any) *ConfigDefinition {
	// Validate cross-language compatibility
	for _, tier := range []struct {
		name   string
		schema map[string]any
	}{
		{"public", publicSchema},
		{"secret", secretSchema},
		{"feature_flags", featureFlagSchema},
	} {
		if tier.schema != nil {
			result := ValidateSmooaiSchema(tier.schema)
			if !result.Valid {
				for _, e := range result.Errors {
					fmt.Fprintf(os.Stderr, "[Smooai Config] Warning: [%s] %s: %s Suggestion: %s\n",
						tier.name, e.Path, e.Message, e.Suggestion)
				}
			}
		}
	}

	emptyObj := map[string]any{"type": "object", "properties": map[string]any{}}

	public := publicSchema
	if public == nil {
		public = map[string]any{}
	}
	secret := secretSchema
	if secret == nil {
		secret = map[string]any{}
	}
	flags := featureFlagSchema
	if flags == nil {
		flags = map[string]any{}
	}

	publicProp := publicSchema
	if publicProp == nil {
		publicProp = emptyObj
	}
	secretProp := secretSchema
	if secretProp == nil {
		secretProp = emptyObj
	}
	flagsProp := featureFlagSchema
	if flagsProp == nil {
		flagsProp = emptyObj
	}

	jsonSchema := map[string]any{
		"$schema": "https://json-schema.org/draft/2020-12/schema",
		"type":    "object",
		"properties": map[string]any{
			"public":        publicProp,
			"secret":        secretProp,
			"feature_flags": flagsProp,
		},
	}

	return &ConfigDefinition{
		PublicSchema:      public,
		SecretSchema:      secret,
		FeatureFlagSchema: flags,
		JSONSchema:        jsonSchema,
	}
}

// MarshalJSON implements custom JSON marshaling for ConfigTier.
func (t ConfigTier) MarshalJSON() ([]byte, error) {
	return json.Marshal(string(t))
}

// UnmarshalJSON implements custom JSON unmarshaling for ConfigTier.
func (t *ConfigTier) UnmarshalJSON(data []byte) error {
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}
	*t = ConfigTier(s)
	return nil
}

// IsValid returns true if the tier is one of the known values.
func (t ConfigTier) IsValid() bool {
	switch t {
	case TierPublic, TierSecret, TierFeatureFlag:
		return true
	}
	return false
}

// DefineConfigTyped creates a configuration definition from Go struct types.
// Each parameter should be a pointer to a struct (or nil for empty tiers).
// The struct's JSON schema is generated via reflection using struct tags.
//
// Example:
//
//	type PublicConfig struct {
//	    APIUrl     string `json:"api_url"`
//	    MaxRetries int    `json:"max_retries" jsonschema:"minimum=0"`
//	}
//
//	config, err := DefineConfigTyped(&PublicConfig{}, nil, nil)
func DefineConfigTyped(publicType, secretType, featureFlagType any) (*ConfigDefinition, error) {
	publicSchema, err := reflectSchema(publicType)
	if err != nil {
		return nil, fmt.Errorf("public schema: %w", err)
	}

	secretSchema, err := reflectSchema(secretType)
	if err != nil {
		return nil, fmt.Errorf("secret schema: %w", err)
	}

	featureFlagSchema, err := reflectSchema(featureFlagType)
	if err != nil {
		return nil, fmt.Errorf("feature flag schema: %w", err)
	}

	return DefineConfig(publicSchema, secretSchema, featureFlagSchema), nil
}

// reflectSchema generates a JSON Schema map from a Go struct type.
// Returns nil for nil inputs (empty schema tiers).
//
// The invopop/jsonschema reflector wraps the schema in $ref + $defs.
// This function inlines the top-level $ref to produce a flat schema
// with "type", "properties", "required" etc. at the top level.
func reflectSchema(v any) (map[string]any, error) {
	if v == nil {
		return nil, nil
	}

	r := &jsonschema.Reflector{}
	schema := r.Reflect(v)
	if schema == nil {
		return nil, nil
	}

	// Convert to map[string]any via JSON roundtrip
	data, err := json.Marshal(schema)
	if err != nil {
		return nil, fmt.Errorf("marshal schema: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("unmarshal schema: %w", err)
	}

	// Inline the top-level $ref if present
	result = inlineTopLevelRef(result)

	return result, nil
}

// inlineTopLevelRef resolves a top-level "$ref": "#/$defs/TypeName" by
// replacing the root schema with the referenced definition.
func inlineTopLevelRef(schema map[string]any) map[string]any {
	ref, hasRef := schema["$ref"].(string)
	if !hasRef {
		return schema
	}

	defs, hasDefs := schema["$defs"].(map[string]any)
	if !hasDefs {
		return schema
	}

	// Parse "#/$defs/TypeName"
	const prefix = "#/$defs/"
	if len(ref) <= len(prefix) || ref[:len(prefix)] != prefix {
		return schema
	}
	typeName := ref[len(prefix):]

	def, found := defs[typeName].(map[string]any)
	if !found {
		return schema
	}

	// Build the inlined schema: copy the definition, then add remaining $defs
	// (for schemas that reference other types)
	inlined := make(map[string]any, len(def)+1)
	for k, v := range def {
		inlined[k] = v
	}

	// Copy any remaining $defs (excluding the inlined one) for nested references
	remaining := make(map[string]any)
	for k, v := range defs {
		if k != typeName {
			remaining[k] = v
		}
	}
	if len(remaining) > 0 {
		inlined["$defs"] = remaining
	}

	return inlined
}

File: rust/config/src/schema.rs
//! Configuration schema definition using serde.

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::schema_validator::validate_smooai_schema;

/// Configuration value tiers.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ConfigTier {
    Public,
    Secret,
    FeatureFlag,
}

/// Result of defining a configuration schema.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfigDefinition {
    pub public_schema: serde_json::Value,
    pub secret_schema: serde_json::Value,
    pub feature_flag_schema: serde_json::Value,
    pub json_schema: serde_json::Value,
}

/// Define a configuration schema from JSON schema components.
///
/// Validates each tier's schema for cross-language compatibility. Returns
/// an error if any schema uses unsupported JSON Schema features.
pub fn define_config(
    public_schema: Option<serde_json::Value>,
    secret_schema: Option<serde_json::Value>,
    feature_flag_schema: Option<serde_json::Value>,
) -> ConfigDefinition {
    // Validate cross-language compatibility
    for (tier, schema) in &[
        ("public", &public_schema),
        ("secret", &secret_schema),
        ("feature_flags", &feature_flag_schema),
    ] {
        if let Some(s) = schema {
            let result = validate_smooai_schema(s);
            if !result.valid {
                let msgs: Vec<String> = result
                    .errors
                    .iter()
                    .map(|e| format!("  {}: {} Suggestion: {}", e.path, e.message, e.suggestion))
                    .collect();
                eprintln!(
                    "[Smooai Config] Warning: [{}] Schema uses unsupported features:\n{}",
                    tier,
                    msgs.join("\n")
                );
            }
        }
    }

    let empty_obj = serde_json::json!({"type": "object", "properties": {}});

    let public = public_schema.clone().unwrap_or_default();
    let secret = secret_schema.clone().unwrap_or_default();
    let feature_flags = feature_flag_schema.clone().unwrap_or_default();

    let json_schema = serde_json::json!({
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "type": "object",
        "properties": {
            "public": public_schema.unwrap_or(empty_obj.clone()),
            "secret": secret_schema.unwrap_or(empty_obj.clone()),
            "feature_flags": feature_flag_schema.unwrap_or(empty_obj),
        }
    });

    ConfigDefinition {
        public_schema: public,
        secret_schema: secret,
        feature_flag_schema: feature_flags,
        json_schema,
    }
}

/// Marker struct for empty schema tiers when using `define_config_typed`.
///
/// Use this for tiers that have no configuration values.
///
/// ```
/// use smooai_config::schema::{define_config_typed, EmptySchema};
/// use schemars::JsonSchema;
/// use serde::{Serialize, Deserialize};
///
/// #[derive(Default, Serialize, Deserialize, JsonSchema)]
/// struct MyPublicConfig {
///     api_url: String,
///     max_retries: u32,
/// }
///
/// let config = define_config_typed::<MyPublicConfig, EmptySchema, EmptySchema>();
/// ```
#[derive(Default, Serialize, Deserialize, JsonSchema)]
pub struct EmptySchema {}

/// Define a configuration schema using native Rust types that implement `JsonSchema`.
///
/// This is the preferred way to define config schemas in Rust. The struct types
/// are automatically converted to JSON Schema via `schemars`, then validated
/// for cross-language compatibility before being passed to `define_config()`.
///
/// Use `EmptySchema` for tiers that have no configuration values.
pub fn define_config_typed<P, S, F>() -> ConfigDefinition
where
    P: JsonSchema + Default,
    S: JsonSchema + Default,
    F: JsonSchema + Default,
{
    let public_schema = schema_or_none::<P>();
    let secret_schema = schema_or_none::<S>();
    let feature_flag_schema = schema_or_none::<F>();

    define_config(public_schema, secret_schema, feature_flag_schema)
}

/// Convert a `JsonSchema` type to a `serde_json::Value`, returning `None` for `EmptySchema`.
fn schema_or_none<T: JsonSchema>() -> Option<serde_json::Value> {
    // Check if this is EmptySchema by looking at the generated schema
    let schema = schemars::schema_for!(T);
    let value = serde_json::to_value(&schema).unwrap_or_default();

    // EmptySchema generates a schema with no properties (or empty object)
    // Check if this is effectively empty
    if std::any::type_name::<T>().ends_with("EmptySchema") {
        return None;
    }

    Some(value)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_config() {
        let result = define_config(None, None, None);
        assert_eq!(result.public_schema, serde_json::Value::default());
        assert_eq!(result.secret_schema, serde_json::Value::default());
        assert_eq!(result.feature_flag_schema, serde_json::Value::default());
        assert_eq!(result.json_schema["type"], "object");
    }

    #[test]
    fn test_empty_config_has_json_schema_draft() {
        let result = define_config(None, None, None);
        assert_eq!(
            result.json_schema["$schema"],
            "https://json-schema.org/draft/2020-12/schema"
        );
    }

    #[test]
    fn test_empty_config_has_all_tier_properties() {
        let result = define_config(None, None, None);
        let props = &result.json_schema["properties"];
        assert!(props["public"].is_object());
        assert!(props["secret"].is_object());
        assert!(props["feature_flags"].is_object());
    }

    #[test]
    fn test_with_public_schema() {
        let public = serde_json::json!({
            "type": "object",
            "properties": {
                "api_url": {"type": "string"},
                "max_retries": {"type": "integer"}
            }
        });
        let result = define_config(Some(public.clone()), None, None);
        assert_eq!(result.public_schema, public);
        assert_eq!(result.json_schema["properties"]["public"], public);
    }

    #[test]
    fn test_with_secret_schema() {
        let secret = serde_json::json!({
            "type": "object",
            "properties": {
                "api_key": {"type": "string"},
                "jwt_secret": {"type": "string"}
            }
        });
        let result = define_config(None, Some(secret.clone()), None);
        assert_eq!(result.secret_schema, secret);
        assert_eq!(result.json_schema["properties"]["secret"], secret);
        // Public should be empty default
        assert_eq!(result.json_schema["properties"]["public"]["type"], "object");
    }

    #[test]
    fn test_with_feature_flag_schema() {
        let flags = serde_json::json!({
            "type": "object",
            "properties": {
                "enable_new_ui": {"type": "boolean"},
                "beta_features": {"type": "boolean"}
            }
        });
        let result = define_config(None, None, Some(flags.clone()));
        assert_eq!(result.feature_flag_schema, flags);
        assert_eq!(result.json_schema["properties"]["feature_flags"], flags);
    }

    #[test]
    fn test_with_all_tiers() {
        let public = serde_json::json!({"type": "object", "properties": {"url": {"type": "string"}}});
        let secret = serde_json::json!({"type": "object", "properties": {"key": {"type": "string"}}});
        let flags = serde_json::json!({"type": "object", "properties": {"beta": {"type": "boolean"}}});

        let result = define_config(Some(public.clone()), Some(secret.clone()), Some(flags.clone()));
        assert_eq!(result.public_schema, public);
        assert_eq!(result.secret_schema, secret);
        assert_eq!(result.feature_flag_schema, flags);
        assert_eq!(result.json_schema["properties"]["public"], public);
        assert_eq!(result.json_schema["properties"]["secret"], secret);
        assert_eq!(result.json_schema["properties"]["feature_flags"], flags);
    }

    #[test]
    fn test_config_tier_serialization() {
        let tier = ConfigTier::FeatureFlag;
        let json = serde_json::to_string(&tier).unwrap();
        assert_eq!(json, "\"feature_flag\"");
    }

    #[test]
    fn test_config_tier_public_serialization() {
        let json = serde_json::to_string(&ConfigTier::Public).unwrap();
        assert_eq!(json, "\"public\"");
    }

    #[test]
    fn test_config_tier_secret_serialization() {
        let json = serde_json::to_string(&ConfigTier::Secret).unwrap();
        assert_eq!(json, "\"secret\"");
    }

    #[test]
    fn test_config_tier_deserialization() {
        let public: ConfigTier = serde_json::from_str("\"public\"").unwrap();
        assert_eq!(public, ConfigTier::Public);

        let secret: ConfigTier = serde_json::from_str("\"secret\"").unwrap();
        assert_eq!(secret, ConfigTier::Secret);

        let flag: ConfigTier = serde_json::from_str("\"feature_flag\"").unwrap();
        assert_eq!(flag, ConfigTier::FeatureFlag);
    }

    #[test]
    fn test_config_tier_invalid_deserialization() {
        let result: Result<ConfigTier, _> = serde_json::from_str("\"invalid\"");
        assert!(result.is_err());
    }

    #[test]
    fn test_config_definition_serialization_roundtrip() {
        let public = serde_json::json!({"type": "object", "properties": {"url": {"type": "string"}}});
        let result = define_config(Some(public), None, None);
        let json = serde_json::to_string(&result).unwrap();
        let deserialized: ConfigDefinition = serde_json::from_str(&json).unwrap();
        assert_eq!(result.public_schema, deserialized.public_schema);
        assert_eq!(result.json_schema, deserialized.json_schema);
    }

    #[test]
    fn test_complex_nested_schema() {
        let public = serde_json::json!({
            "type": "object",
            "properties": {
                "database": {
                    "type": "object",
                    "properties": {
                        "host": {"type": "string", "default": "localhost"},
                        "port": {"type": "integer", "default": 5432},
                        "options": {
                            "type": "object",
                            "properties": {
                                "ssl": {"type": "boolean"},
                                "pool_size": {"type": "integer"}
                            }
                        }
                    }
                }
            }
        });
        let result = define_config(Some(public.clone()), None, None);
        assert_eq!(
            result.json_schema["properties"]["public"]["properties"]["database"]["properties"]["host"]["type"],
            "string"
        );
    }

    // --- define_config_typed tests ---

    #[derive(Default, serde::Serialize, serde::Deserialize, schemars::JsonSchema)]
    struct TestPublicConfig {
        api_url: String,
        max_retries: u32,
        debug: bool,
    }

    #[derive(Default, serde::Serialize, serde::Deserialize, schemars::JsonSchema)]
    struct TestSecretConfig {
        api_key: String,
    }

    #[derive(Default, serde::Serialize, serde::Deserialize, schemars::JsonSchema)]
    struct TestFeatureFlags {
        enable_beta: bool,
    }

    #[test]
    fn test_define_config_typed_basic() {
        let result = define_config_typed::<TestPublicConfig, EmptySchema, EmptySchema>();
        // Public schema should have properties from TestPublicConfig
        let public = &result.json_schema["properties"]["public"];
        assert!(public.is_object());
        let props = &public["properties"];
        assert!(props["api_url"].is_object());
        assert!(props["max_retries"].is_object());
        assert!(props["debug"].is_object());
    }

    #[test]
    fn test_define_config_typed_all_tiers() {
        let result = define_config_typed::<TestPublicConfig, TestSecretConfig, TestFeatureFlags>();
        let public = &result.json_schema["properties"]["public"];
        let secret = &result.json_schema["properties"]["secret"];
        let flags = &result.json_schema["properties"]["feature_flags"];
        assert!(public["properties"]["api_url"].is_object());
        assert!(secret["properties"]["api_key"].is_object());
        assert!(flags["properties"]["enable_beta"].is_object());
    }

    #[test]
    fn test_define_config_typed_empty_schema() {
        let result = define_config_typed::<EmptySchema, EmptySchema, EmptySchema>();
        // Should produce valid config with empty tiers
        assert_eq!(result.json_schema["type"], "object");
    }
}
